client/python/gradio_client/client.py

```diff
@@ -0,0 +1,298 @@
+                headers=self.headers,
+            )
+            .json()
+            .get("host")
+        )
+
+    def _get_config(self) -> Dict:
+        assert self.src is not None
+        r = requests.get(self.src, headers=self.headers)
+        # some basic regex to extract the config
+        result = re.search(r"window.gradio_config = (.*?);[\s]*</script>", r.text)
+        try:
+            config = json.loads(result.group(1))  # type: ignore
+        except AttributeError:
+            raise ValueError(f"Could not get Gradio config from: {self.src}")
+        if "allow_flagging" in config:
+            raise ValueError(
+                f"Gradio 2.x is not supported by this client. Please upgrade this app to Gradio 3.x."
+            )
+        return config
+
+
+class Endpoint:
+    """Helper class for storing all the information about a single API endpoint."""
+
+    def __init__(self, client: Client, fn_index: int, dependency: Dict):
+        self.api_url = client.api_url
+        self.ws_url = client.ws_url
+        self.fn_index = fn_index
+        self.dependency = dependency
+        self.headers = client.headers
+        self.config = client.config
+        self.use_ws = self._use_websocket(self.dependency)
+        self.hf_token = client.hf_token
+        try:
+            self.serializers, self.deserializers = self._setup_serializers()
+            self.is_valid = self.dependency[
+                "backend_fn"
+            ]  # Only a real API endpoint if backend_fn is True
+        except AssertionError:
+            self.is_valid = False
+
+    def end_to_end_fn(self, *data):
+        if not self.is_valid:
+            raise utils.InvalidAPIEndpointError()
+        inputs = self.serialize(*data)
+        predictions = self.predict(*inputs)
+        outputs = self.deserialize(*predictions)
+        if len(self.dependency["outputs"]) == 1:
+            return outputs[0]
+        return outputs
+
+    def predict(self, *data) -> Tuple:
+        data = json.dumps({"data": data, "fn_index": self.fn_index})
+        hash_data = json.dumps(
+            {"fn_index": self.fn_index, "session_hash": str(uuid.uuid4())}
+        )
+        if self.use_ws:
+            result = utils.synchronize_async(self._ws_fn, data, hash_data)
+            output = result["data"]
+        else:
+            response = requests.post(self.api_url, headers=self.headers, data=data)
+            result = json.loads(response.content.decode("utf-8"))
+            try:
+                output = result["data"]
+            except KeyError:
+                if "error" in result and "429" in result["error"]:
+                    raise utils.TooManyRequestsError(
+                        "Too many requests to the Hugging Face API"
+                    )
+                raise KeyError(
+                    f"Could not find 'data' key in response. Response received: {result}"
+                )
+        return tuple(output)
```

Comment:
Curious how come we're returning tuples everywhere given that "data" is already a list?

Reply:
It helped with the typing I believe